rtl/include/eclhelper.hpp

```diff
@@ -780,8 +780,312 @@ interface ICodeContext : public IResourceContext
     virtual IEngineRowAllocator * getRowAllocatorEx(IOutputMetaData * meta, unsigned activityId, unsigned flags) const = 0;
     virtual void addWuExceptionEx(const char * text, unsigned code, unsigned severity, unsigned audience, const char * source) = 0;
     virtual unsigned getElapsedMs() const = 0;
+    virtual unsigned getWorkflowId() const = 0; // Note: don't use yet as it has not been fully implemented in all derived classes
 };
 
+class IndirectCodeContext : implements ICodeContext
```

Comment:
These classes need to be in a different file - this is used from generated code, and should contain the minimum possible - especially no implementations.

There is also a second danger of including class implementations in a header.  If a compiler only sees one concrete implementation of a class it will add code to de-virtualise the function calls (possibly only if it has an inline function definition, I have forgotten!).  That does not work well if the implementation is in a different dll, and also is counter-productive if it is not the most common implementation.

Having said that it isn't clear where this class should go if it is being called from the workflow.  I think the correct answer may be that the code in workflow.cpp needs to split in two, the workflowMachine should be moved to throhelper, and this should live in thorhelper as it did originally.

A less painful solution would be to add a new header/c++ file to eclrtl, but move it out of this file.